src/ansible/provisioning.rs

```diff
@@ -0,0 +1,384 @@
+    extra_vars::ExtraVarsDocBuilder, AnsibleInventoryType, AnsiblePlaybook, AnsibleRunner,
+};
+use crate::{
+    deploy::DeployOptions,
+    error::{Error, Result},
+    print_duration, BinaryOption, CloudProvider, LogFormat, SshClient, UpgradeOptions,
+};
+use semver::Version;
+use std::{
+    net::{IpAddr, SocketAddr},
+    time::Instant,
+};
+
+const DEFAULT_BETA_ENCRYPTION_KEY: &str =
+    "49113d2083f57a976076adbe85decb75115820de1e6e74b47e0429338cef124a";
+
+pub enum NodeType {
+    Bootstrap,
+    Normal,
+}
+
+pub struct ProvisionOptions {
+    pub beta_encryption_key: Option<String>,
+    pub binary_option: BinaryOption,
+    pub bootstrap_node_count: u16,
+    pub env_variables: Option<Vec<(String, String)>>,
+    pub log_format: Option<LogFormat>,
+    pub logstash_details: Option<(String, Vec<SocketAddr>)>,
+    pub name: String,
+    pub node_count: u16,
+    pub public_rpc: bool,
+}
+
+impl From<DeployOptions> for ProvisionOptions {
+    fn from(deploy_options: DeployOptions) -> Self {
+        ProvisionOptions {
+            beta_encryption_key: deploy_options.beta_encryption_key,
+            binary_option: deploy_options.binary_option,
+            bootstrap_node_count: deploy_options.bootstrap_node_count,
+            env_variables: deploy_options.env_variables,
+            log_format: deploy_options.log_format,
+            logstash_details: deploy_options.logstash_details,
+            name: deploy_options.name,
+            node_count: deploy_options.node_count,
+            public_rpc: deploy_options.public_rpc,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct AnsibleProvisioner {
+    pub ansible_runner: AnsibleRunner,
+    pub cloud_provider: CloudProvider,
+    pub ssh_client: SshClient,
+}
+
+impl AnsibleProvisioner {
+    pub fn new(
+        ansible_runner: AnsibleRunner,
+        cloud_provider: CloudProvider,
+        ssh_client: SshClient,
+    ) -> Self {
+        Self {
+            ansible_runner,
+            cloud_provider,
+            ssh_client,
+        }
+    }
+
+    pub async fn build_safe_network_binaries(&self, options: &ProvisionOptions) -> Result<()> {
+        let start = Instant::now();
+        println!("Obtaining IP address for build VM...");
+        let build_inventory = self
+            .ansible_runner
+            .get_inventory(AnsibleInventoryType::Build, true)
+            .await?;
+        let build_ip = build_inventory[0].1;
+        self.ssh_client
+            .wait_for_ssh_availability(&build_ip, &self.cloud_provider.get_ssh_user())?;
+
+        println!("Running ansible against build VM...");
+        let extra_vars = self.build_binaries_extra_vars_doc(options)?;
+        self.ansible_runner.run_playbook(
+            AnsiblePlaybook::Build,
+            AnsibleInventoryType::Build,
+            Some(extra_vars),
+        )?;
+        print_duration(start.elapsed());
+        Ok(())
+    }
+
+    pub async fn copy_logs(&self, name: &str, resources_only: bool) -> Result<()> {
```

Comment:
Should we also fetch the logs of the bootstrap nodes here?

Reply:
Yeah, I'm not sure if anybody is using this for logs any more, but yeah, it can be extended for the bootstrap nodes.